.github/workflows/publish.yaml

```diff
@@ -7,7 +7,50 @@ on:
       - "*"
 
 jobs:
+  test-java-compatibility:
+    name: Test Java Compatibility
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        java-version: [8, 11, 17, 21]
+    defaults:
+      run:
+        working-directory: ./client/java
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Set up JDK ${{ matrix.java-version }}
+        uses: actions/setup-java@v4
+        with:
+          java-version: ${{ matrix.java-version }}
+          distribution: 'temurin'
+
+      - name: Cache Maven dependencies
+        uses: actions/cache@v4
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-java${{ matrix.java-version }}-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-m2-java${{ matrix.java-version }}-
+            ${{ runner.os }}-m2-
+
+      - name: Generate protobuf sources
+        run: mvn protobuf:compile protobuf:compile-custom
+
+      - name: Compile Kotlin and Java sources
+        run: mvn compile
+
+      - name: Run tests
+        run: mvn test
+
+      - name: Verify JAR can be built
+        run: mvn package -DskipTests
```

Comment:
This should not be in the publish workflow (we've already created release tags at the point its practically too late.

This test should block code from being integrated into main, so should be for PRs/branches and be run conditionally on changes to this SDK.